frame/elections-phragmen/src/lib.rs

```diff
@@ -633,116 +698,160 @@ decl_module! {
 	///
-	/// If a replacement exists, `Members` and `RunnersUp` storage is updated, where the first
-	/// element of `RunnersUp` is used as the replacement and `Ok(true)` is returned. Else,
-	/// `Ok(false)` is returned with no storage updated.
+	/// Both `Members` and `RunnersUp` storage is updated accordingly. `T::ChangeMember` is called
+	/// if needed. If `slash` is true, the deposit of the potentially removed member is slashed,
+	/// else it is unreserved.
 	///
-	/// Note that this function _will_ call into `T::ChangeMembers` in case any change happens
-	/// (`Ok(true)`).
+	/// ### Note
 	///
-	/// If replacement exists, this will read and write from/into both `Members` and `RunnersUp`.
-	fn remove_and_replace_member(who: &T::AccountId) -> Result<bool, DispatchError> {
-		let mut members_with_stake = Self::members();
-		if let Ok(index) = members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(who)) {
-			members_with_stake.remove(index);
-
-			let next_up = <RunnersUp<T>>::mutate(|runners_up| runners_up.pop());
-			let maybe_replacement = next_up.and_then(|(replacement, stake)|
-				members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(&replacement))
+	/// This function attempts to preserve the prime. If the removed members is not the prime, it is
+	/// set again via [`Trait::ChangeMembers` ] to prevent it being wiped.
+	fn remove_and_replace_member(who: &T::AccountId, slash: bool) -> Result<bool, DispatchError> {
+		// closure will return:
+		// - `Ok(Option(replacement))` if member was removed and replacement was replaced.
+		// - `Ok(None)` if member was removed but no replacement was found
+		// - `Err(_)` if who is not a member.
+		<Members<T>>::try_mutate(|members| {
+			// we remove the member anyhow, regardless of having a runner-up or not.
+			// TODO: a way to make sure all tests operate in a way that accounts are not sorted.
+			let remove_index = members
+				// note: members are always sorted.
+				.binary_search_by(|m| m.who.cmp(who))
+				.map_err(|_| Error::<T>::NotMember)?;
+			let removed = members.remove(remove_index);
+			if slash {
+				let (imbalance, _remaining) = T::Currency::slash_reserved(who, removed.deposit);
+				debug_assert_eq!(_remaining, 0u32.into());
+				T::LoserCandidate::on_unbalanced(imbalance);
+				Self::deposit_event(RawEvent::SeatHolderSlashed(who.clone(), removed.deposit));
+			} else {
+				T::Currency::unreserve(who, removed.deposit);
+			}
+
+			Ok(<RunnersUp<T>>::mutate(|r| r.pop()).map(|next_best| {
+				// invariant: Members and runners-up are disjoint. This will always be err and give
+				// us an index to insert.
+				members
+					.binary_search_by(|m| m.who.cmp(&next_best.who))
 					.err()
 					.map(|index| {
-						members_with_stake.insert(index, (replacement.clone(), stake));
-						replacement
-					})
-			);
+						members.insert(index, next_best.clone());
+					});
+				next_best
+			}))
+		})
+		.map(|maybe_replacement| {
+			let member_ids = Self::members()
```

Comment:
maybe a nit, but here you query the overlay again, when def you could probably keep the members list in memory